ably/realtime_channel.go

```diff
@@ -2,130 +2,342 @@ package ably
+	}
 	msg := &proto.ProtocolMessage{
-		Action:  proto.ActionMessage,
-		Channel: c.Name,
-		Messages: []*proto.Message{
-			{Name: name, Data: data},
-		},
+		Action:   proto.ActionMessage,
+		Channel:  c.Name,
+		Messages: messages,
 	}
-	return c.client.send(msg)
+	switch c.State() {
+	case StateChanInitialized, StateChanAttaching:
+		if c.client.opts.NoQueueing {
+			return errQueueing
+		}
+		c.queue.Enqueue(msg, listen)
+		return nil
+	case StateChanAttached:
+	default:
+		return &Error{Code: 90001}
+	}
+	return c.client.Connection.send(msg, listen)
+}
+
+// State gives current state of the channel.
+func (c *RealtimeChannel) State() int {
+	c.state.Lock()
+	state := c.state.current
+	c.state.Unlock()
+	return state
+}
+
+// Reason gives the last error that caused channel transition to failed state.
+func (c *RealtimeChannel) Reason() error {
+	c.state.Lock()
+	err := c.state.err
+	c.state.Unlock()
+	return err
 }
 
 func (c *RealtimeChannel) notify(msg *proto.ProtocolMessage) {
 	switch msg.Action {
 	case proto.ActionAttached:
-		c.setState(ChanStateAttached)
+		c.state.syncSet(StateChanAttached, nil)
+		c.queue.Flush()
 	case proto.ActionDetached:
-		c.setState(ChanStateDetached)
+		c.state.syncSet(StateChanDetached, nil)
 	case proto.ActionPresence:
-		// TODO
+		Log.Printf(LogError, "TODO: implement RealtimePresence: %v", msg)
 	case proto.ActionError:
-		c.setState(ChanStateFailed)
-		c.Err = msg.Error
-		// TODO c.Close()
+		c.state.syncSet(StateChanFailed, newErrorProto(msg.Error))
+		c.queue.Fail(newErrorProto(msg.Error))
+		// TODO recovery
 	case proto.ActionMessage:
-		c.listenMtx.RLock()
-		defer c.listenMtx.RUnlock()
-		for _, m := range msg.Messages {
-			if l, ok := c.listeners[""]; ok {
+		c.listenMtx.Lock()
+		for _, msg := range msg.Messages {
+			if l, ok := c.listen[""]; ok {
 				for ch := range l {
-					ch <- m
+					select {
+					case ch <- msg:
+					default:
+						Log.Printf(LogWarn, "dropped %q message due to slow receiver", msg.Name)
+					}
 				}
 			}
-			if l, ok := c.listeners[m.Name]; ok {
+			if l, ok := c.listen[msg.Name]; ok {
 				for ch := range l {
-					ch <- m
+					select {
+					case ch <- msg:
+					default:
+						Log.Printf(LogWarn, "dropped %q message due to slow receiver", msg.Name)
+					}
 				}
 			}
 		}
+		c.listenMtx.Unlock()
 	default:
 	}
 }
 
-func (c *RealtimeChannel) Attach() {
-	c.stateMtx.Lock()
-	defer c.stateMtx.Unlock()
-	if c.State == ChanStateAttaching || c.State == ChanStateAttached {
-		return
-	}
-	if !c.client.isActive() {
-		c.Err = errors.New("Connection not active")
-		return
-	}
-	msg := &proto.ProtocolMessage{Action: proto.ActionAttach, Channel: c.Name}
-	if err := c.client.send(msg); err != nil {
-		c.Err = fmt.Errorf("Attach request failed: %s", err)
-		return
-	}
-	c.State = ChanStateAttaching
-}
-
-func (c *RealtimeChannel) setState(s ChanState) {
-	c.stateMtx.Lock()
-	defer c.stateMtx.Unlock()
-	c.State = s
+func (c *RealtimeChannel) isActive() bool {
```

Comment:
Reading the code, I understand why `isActive` doesn't use a lock here since you are managing the lock outside of this method. But surely, you should be consistent with the other `isActive` on the connection too and not use locks.